src/infra/core/src/repos/dataset_impl.rs

```diff
@@ -133,74 +133,80 @@ where
         current_head: &Multihash,
         last_seen: Option<&Multihash>,
     ) -> Result<UpdateSummaryIncrement, GetSummaryError> {
-        use tokio_stream::StreamExt;
-
-        let mut block_stream =
-            self.metadata_chain
-                .iter_blocks_interval(current_head, last_seen, true);
-        let mut increment = UpdateSummaryIncrement {
-            seen_head: Some(current_head.clone()),
-            ..Default::default()
-        };
-
-        while let Some((_, block)) = block_stream.try_next().await.int_err()? {
-            match block.event {
-                MetadataEvent::Seed(seed) => {
-                    increment.seen_id.get_or_insert(seed.dataset_id);
-                    increment.seen_kind.get_or_insert(seed.dataset_kind);
-                }
-                MetadataEvent::SetTransform(set_transform) => {
-                    if increment.seen_dependencies.is_none() {
-                        increment.seen_dependencies = Some(
-                            set_transform
-                                .inputs
-                                .into_iter()
-                                .map(|i| i.dataset_ref.id().cloned().unwrap())
-                                .collect(),
-                        );
-                    }
-                }
-                MetadataEvent::AddData(add_data) => {
-                    increment.seen_last_pulled.get_or_insert(block.system_time);
-
-                    if let Some(output_data) = add_data.new_data {
-                        let iv = output_data.offset_interval;
-                        increment.seen_num_records += iv.end - iv.start + 1;
+        type Flag = MetadataEventTypeFlags;
+        type Decision = MetadataVisitorDecision;
 
-                        increment.seen_data_size += output_data.size;
-                    }
+        struct VisitorState {
+            requested_flags: Flag,
+            increment: UpdateSummaryIncrement,
+        }
 
-                    if let Some(checkpoint) = add_data.new_checkpoint {
-                        increment.seen_checkpoints_size += checkpoint.size;
+        const INITIAL_REQUESTED_FLAGS: Flag = Flag::SEED
+            .union(Flag::SET_TRANSFORM)
+            .union(Flag::DATA_BLOCK);
```

Comment:
This is another case where I think `iter_blocks_interval` is simpler and as fast. Summaries need all data events so they won't be able to skip any blocks.

Reply:
Reverted -- Done